pipeline/execout/cachev1/engine.go

```diff
@@ -0,0 +1,143 @@
+	}
+	return e, nil
+}
+
+func (e *Engine) Prime(modules map[string]*pbsubstreams.Module) error {
+	for hash, module := range modules {
+		if err := e.registerCache(module, hash); err != nil {
+			return fmt.Errorf("failed to register chache for module %q: %w", module.Name, err)
+		}
+	}
+	return nil
+}
+
+func (e *Engine) FlushAndUpdate(ctx context.Context, blockRef bstream.BlockRef) error {
+	e.logger.Info("flushing output caches and updating")
+	for _, moduleCache := range e.caches {
+		if moduleCache.c.IsOutOfRange(blockRef) {
+			if err := e.flushCache(ctx, moduleCache.c); err != nil {
+				return fmt.Errorf("failed to flush cache: %w", err)
+			}
+
+			if _, err := moduleCache.c.LoadAtBlock(ctx, moduleCache.c.CurrentBlockRange.ExclusiveEndBlock); err != nil {
+				return fmt.Errorf("loading blocks %d for module kv %s: %w", moduleCache.c.CurrentBlockRange.ExclusiveEndBlock, moduleCache.c.ModuleName, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *Engine) Flush(ctx context.Context) error {
+	e.logger.Info("flushing output caches")
+	for _, moduleCache := range e.caches {
+		if err := e.flushCache(ctx, moduleCache.c); err != nil {
+			return fmt.Errorf("failed to flush cache: %w", err)
+		}
+	}
+	return nil
+}
+
+func (e *Engine) UndoBlock(moduleName string, blockID string) {
+	if outputCache, found := e.caches[moduleName]; found {
+		outputCache.c.Delete(blockID)
+	}
+}
+
+func (e *Engine) flushCache(ctx context.Context, cache *OutputCache) error {
+	filename := cache.currentFilename()
+	e.logger.Debug("saving cache",
+		zap.String("module_name", cache.ModuleName),
+		zap.Uint64("start_block", cache.CurrentBlockRange.StartBlock),
+		zap.Uint64("end_block", cache.CurrentBlockRange.ExclusiveEndBlock),
+	)
+
+	if err := cache.save(ctx, filename); err != nil {
+		return fmt.Errorf("save: saving outpust or module kv %s: %w", cache.ModuleName, err)
+	}
+	return nil
+}
+
+func (e *Engine) registerCache(module *pbsubstreams.Module, hash string) error {
+	e.logger.Debug("registering modules", zap.String("module_name", module.Name))
+
+	if _, found := e.caches[module.Name]; found {
+		return nil
+	}
+
+	moduleStore, err := e.baseCacheStore.SubStore(fmt.Sprintf("%s/outputs", hash))
+	if err != nil {
+		return fmt.Errorf("failed createing substore: %w", err)
+	}
+
+	e.caches[module.Name] = &OutputCacheState{
+		c:           NewOutputCache(module.Name, moduleStore, e.SaveBlockInterval, e.logger),
+		initialized: false,
+	}
+	return nil
+}
+
+func (e *Engine) get(moduleName string, clock *pbsubstreams.Clock) ([]byte, bool, error) {
+	cache, found := e.caches[moduleName]
+	if !found {
+		return nil, false, fmt.Errorf("cache %q not found", moduleName)
+	}
+	if !cache.initialized {
+		// TODO: should context be passed upstream?
+		if _, err := cache.c.LoadAtBlock(e.ctx, clock.Number); err != nil {
+			return nil, false, fmt.Errorf("unable to load cache %q at block %d: %w", moduleName, clock.Number, err)
+		}
+		cache.initialized = true
+	}
```

Comment:
we could decide to have the `initialized` flag set on the `Engine`, and loop through all of the uninitialized _caches_ and `LoadAtBlock` them in parallel.

The `ctx` for the caching engine ought to come from the top-most initialization of the process, shouldn't it? If we Ctrl+C, then perhaps we cancel the loads and saves, not based on the requests' contexts.